src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5440,20 +5441,37 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
             LearnFromNonNullTest(leftOperand, ref leftState);
             LearnFromNullTest(leftOperand, ref this.State);
 
+            var adjustedNodeType = node.Type;
+
+            if (node.Type?.IsDynamic() == true && leftOperand is BoundIndexerAccess { Type.TypeKind: not TypeKind.Dynamic } indexerAccess)
```

Comment:
>node.Type?.IsDynamic() == true && leftOperand is BoundIndexerAccess { Type.TypeKind: not TypeKind.Dynamic } indexerAccess [](http://example.com/codeflow?start=16&length=121)

Consider extracting the various checks that follow this recurring pattern into one helper. That would help clarify the intent (detect a statically-bound indexer access with dynamic arguments in an assignment node) and offers a place to leave a comment and tie the various places that do similar checks together.